### PR TITLE
Destroying a DisplayObject will remove itself

### PR DESCRIPTION
**Pull request by bigtimebuddy**
Child removes itself upon being destroyed. 
Responds to issue #2800

### CODE REVIEW

**Comment by ivanpopelyshev** [comment]
NONONONONONO WHYYY!

Do you realize what will happen in container.destroy(true) ?

**Comment by GoodBoyDigital** [comment]
interesting conundrum that!

File: src/core/display/Container.js [hunk 1]
```diff
@@ -572,16 +572,16 @@ Container.prototype.destroy = function (options)
 {
     DisplayObject.prototype.destroy.call(this);
 
+    this.removeChildren();
```
Comment (GoodBoyDigital) [inline src/core/display/Container.js]: unless i'm mistaken - would this line make ```this.children``` empty?

**Comment by bigtimebuddy** [comment]
DOH! You're right. Destroying the children in reverse order should be fine.

**Comment by ivanpopelyshev** [comment]
And now, gentlemans, we have destroy() method that works O(N^2). Nice :)

**Comment by bigtimebuddy** [comment]
If someone wants to implement another approach, I'm cool with that. I think this functionality is important though and would like to see it in v4.

**Comment by ivanpopelyshev** [comment]
I think its better to leave it as it is for now. There are also things like "remove" event in child, and at the same time we dont have "parentDetached" or anything like that... 

The thing is, any small change may break some code, its better if we make big PR that deals with that.

**Comment by bigtimebuddy** [comment]
@ivanpopelyshev I disagree that we should leave it. There's never a case where you'd want to destroy and not remove that child from it's display list. Expecting developers to remember that is a poor design choice.

**Comment by ivanpopelyshev** [comment]
Then destroying whole subtree must be separated to different method, like we have with renderWebGL and _renderWebGL. We still don't have a convention which events must be fired in case of whole subtree is removed, or whole subtree is destroyed.

**Comment by ivanpopelyshev** [comment]
Hm. Yes, good idea: we need to separate destroying this element from destroying whole subtree, also remove all children before we ask their destruction, something like that.

**Comment by GoodBoyDigital** [comment]
Interesting debate going on here!

 I think @bigtimebuddy makes a good point about removing an object from its parent if it has been destroyed. If its going to be destroyed we know it will throw errors if its not removed. @ivanpopelyshev - what kind of use cases are you thinking about where this is undesirable behaviour?

Thanks!

**Comment by ivanpopelyshev** [comment]
Mass destruction. We have to separate methods for killing only this element and killing whole subtree.

**Comment by GoodBoyDigital** [comment]
I see what ya mean, thats a different issue than removing children though?

**Comment by ivanpopelyshev** [comment]
No, it's not. You cant just add "removeChild" in destroy() because of performance reasons in mass destroy.  Its either big hack, either architecture improvement.

**Comment by GoodBoyDigital** [comment]
is removeChild really expensive?

**Comment by bigtimebuddy** [comment]
Maybe Container should removeChildren before it tries to destroy them. Caching the children list before removing wouldn't create so many removeChild calls within DisplayObject.

**Comment by ivanpopelyshev** [comment]
It has both indexOf and removeAt inside. That means its O(N). If we do N removeChilds, its O(N^2). Try 100k bunnies and your CPU will be fried.

**Comment by GoodBoyDigital** [comment]
Thanks for sharing the details @ivanpopelyshev , I guess to solve this issue do we even need to do a recursive remove child operation? We only need to remove the object being destroyed from its parent (rather than all its children)?

**Comment by ivanpopelyshev** [comment]
@GoodBoyDigital 

OK, may be that will work:

```js
//this like bigtimebuddy's
DisplayObject.prototype.destroy = function ()
{
    if (this.parent)
    {
        this.parent.removeChild(this);
    }
    this.transform = null;

    this.parent = null;

    this._bounds = null;
    this._currentBounds = null;
    this._mask = null;

    this.filterArea = null;
};

//this is mine
Container.prototype.destroy = function (options)
{
    DisplayObject.prototype.destroy.call(this);

    var destroyChildren = typeof options === 'boolean' ? options : options && options.children;

    var oldChildren = this.children;
    this.children = null;
   
    if (destroyChildren)
    {
        for (var i = oldChildren.length - 1; i >= 0; i--)
        {
            var child = oldChildren[i];
            child.parent = null;
            child.emit('removed', this);
            child.destroy(options);
        }
    }
};
```

**Comment by bigtimebuddy** [comment]
Updated with @ivanpopelyshev Container destroy.

**Comment by ivanpopelyshev** [comment]
That's a benchmark, but i dont know how to write this. Basically, if it takes more than 1s on any machine, then its wrong. Problem is, its synchronous code:

```js
describe('destroy', function () {
        it('should destroy 100k rabbits without a problem', function () {
            var texture = PIXI.RenderTexture.create(10, 10);
            var topContainer = new PIXI.Container();
            var bottomContainer = new PIXI.Container();

            topContainer.addChild(bottomContainer);
            for (var i=0;i<100000;i++) {
                var sprite = new PIXI.Sprite(texture);
                bottomContainer.addChild(sprite);
            }

            topContainer.destroy({children: true, texture:true});
        });

    });
```

**Comment by GoodBoyDigital** [comment]
One thing - was wandering if we need -> ```child.emit('removed', this);``` ?
Let me know if i'm completely off the mark!

**Comment by ivanpopelyshev** [comment]
I really don't know, its up to you :) But we just had someone ask if "removed" will be called when whole subtree is destroyed, somewhere in the issues, please find him.

**Comment by GoodBoyDigital** [comment]
What do you think @bigtimebuddy ?

**Comment by bigtimebuddy** [comment]
@GoodBoyDigital I think I'm not a big fan of emitting here. When you destroy something, it means that you have no desire to use that thing anymore. If you're still responding to callback, then you clearly should not be destroying it. Instead, I actually would consider `this.removeAllListeners()` in DisplayObject's `destroy`.

**Comment by GoodBoyDigital** [comment]
I like that! Intention is definitely to nuke the object and it would help ensure that it completely free's it from getting caught in events memory!

**Comment by ivanpopelyshev** [comment]
Maybe add special removeChildrenRecursive then?

**Comment by bigtimebuddy** [comment]
Okay, now removes listeners and doesn't emit remove on destroy.

**Comment by ivanpopelyshev** [comment]
Can we write it down somewhere in docs? We really need explanation of how removeChild/destroy works

**Comment by bigtimebuddy** [comment]
Added more documentation for destroy

**Comment by GoodBoyDigital** [comment]
nice work chaps!

**Pull request merged**